### ISSUE TITLE
training darknet with groups

### ISSUE THREAD

**Issue by DRACOyu** [pjreddie/darknet#1150]
i try to use group convolution to speed up the operation，this is my cfg file，this can train ，but When I tested, there was no output。hope to get everyone's help

I added a group to each set of convolutions, I don't know if this is correct.but the size of the model is half of the original，who can help me ，thanks

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
groups=2
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=0
groups=2
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
groups=2
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
groups=2
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
groups=2
activation=leaky

**Comment by alowmen** [comment]
@DRACOyu  hi handsome, I met the same problem.Did you solve this?

**Comment by developer0hye** [comment]
@alowmen hi handsome. Did you solve this?

**Comment by AllenDun** [comment]
there is still no group convolution  or mobilenet implementation in the darknet(yolov3)?

**Comment by kammirzazad** [comment]
@DRACOyu Based on the [source code of conv. layer](https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/src/convolutional_layer.c#L195) I don't think filter groups were properly implemented. For beginners, layer only allocates space for single filter group and not all of them.
It is interesting that later in forward/backward propagation, it is assumed that space is allocated for all groups (see l.nweights/l.groups).

**Comment by JHLee0513** [comment]
@kammirzazad Is there a possible way to fix that? I would really like to get it to work but I have no idea what's going on in the backend 😞

**Comment by chenbohua3** [comment]
@kammirzazad could you please give more detailed clues for the bugs? I have checked the codes about group convolution several times but did not found any problems. I have tested the speed of mobilenetV2 based yolov3 and find the speed is very slow. It takes about 20ms for one depthwise conv layer with cuda and cudnn, but when I switch off cudnn and use cuda only, it only takes 8ms. I do not know where is the problem.

**Comment by kammirzazad** [comment]
@JHLee0513 I was considering using darknet's filter groups for partitioning conv. layers for execution on mobile devices but ended up writing my own conv layer. Therefore, did not attempt to fix darknet but here is what I would do:
1) Modify parser.c::load_convolutional_weights so that it load weights for all groups (drop "l.groups" from "num" assignment)
2) Modify convolutional_layer.c::make_convolutional_layer so that weights are allocated for all groups and nweights is set correctly (drop "l.groups" from "l.weights", "l.weight_updates" and "l.nweights" assignments)
Steps above assume that darknet developers wanted to implement multiple groups using a single conv layer. This is based on the observation that number of the input channels is independent from number of the groups. See [this](https://blog.yani.io/filter-group-tutorial/) for more details.
Note that for filter grouping to work, you need to train with groups in first place i.e. you need weights file for group-trained neural network.

**Comment by kammirzazad** [comment]
@chenbohua3 I am not familiar with mobilenetV2, does it use filter grouping? The bug I was talking about would mainly affect the functionality, so I am not sure how it could change the latency.

**Comment by chenbohua3** [comment]
@kammirzazad I have figured out that the high latency is caused by the wrong conv algorithm choosed by cudnn when there is group convolutions in the networks. 
Yes, MobilenetV2 uses group convolutions in an extreme way, that is, the number of filter groups is equal to the number of input channels.
I still don't understand the bug you talked about above. I have trained MobilenetV2(with group convolutions) in pytorch and import the weights to the darknet framework. Judging from the results, it works normally.

**Comment by kammirzazad** [comment]
@chenbohua3 Can you share your .cfg? Is "groups" parameter set to number of input channels?

**Comment by chenbohua3** [comment]
@kammirzazad Sorry for the late reply, I have not been on github these days:( 
Here is my .cfg  [link](https://www.dropbox.com/s/uhsjznnghchlhqw/m2.cfg?dl=0). Some conv layers of backbone and head structure have "groups" set to number of input channels.

**Comment by kammirzazad** [comment]
@chenbohua3 I see, most likely I misunderstood the implementation of filter groups.

**Comment by chenbohua3** [comment]
@kammirzazad :)

**Comment by cuixing158** [comment]
@chenbohua3  hello, i have seen your mobilenetv2 cfg file, in your file you set "activation=relu6", darknet framework could support this 'relu6' ?

**Comment by chenbohua3** [comment]
@cuixing158 yes, you need add relu6 to the framework (just do like relu which has already implemented in the framework)
As for the speed, it does perform faster than the original one. However, since DW Conv is not optimized well by CuDnn, It is not theoretically fast.
